source/extensions/transport_sockets/tls/ssl_socket.cc

```diff
@@ -60,6 +60,30 @@ void SslSocket::setTransportSocketCallbacks(Network::TransportSocketCallbacks& c
   ASSERT(!callbacks_);
   callbacks_ = &callbacks;
 
+  // Associate this SSL connection with all the certificates (with their potentially different
+  // private key methods).
+  std::vector<Ssl::PrivateKeyMethodProviderSharedPtr> providers =
+      ctx_->getPrivateKeyMethodProviders();
+
+  for (auto const& provider : providers) {
+    Ssl::PrivateKeyConnectionPtr pk_connection =
+        provider->getPrivateKeyConnection(ssl_.get(), *this, callbacks_->connection().dispatcher());
+    // If a private key method provider has been assigned for this certificate, it's a fatal
+    // error if the connection can't be associated with the method.
+    ASSERT(pk_connection);
+    // We keep track of the private key operations for memory management purposes (the operations
+    // object if destroyed when the SslSocket is destroyed). The operations objects are unique
+    // because they need to associated with the SSL objects so that user data can be passed to the
+    // BoringSSL private key methods.
```

Comment:
`SSL` object is passed to each of the `SSL_PRIVATE_KEY_METHOD` methods, so you can access `ex_data` without creating unique object for each connection.

Reply:
Yes. The wording in the comment could be better. There is no "need" of the association, but it is convenient. I'll postpone fixing this a bit if we decide to leave the connection out of the API.